In [1]:
import numpy as np
import pandas as pd

In [2]:
reldir = 'C:\\Users\\shefa\\OneDrive\\CS6220\\msd.recsys\\datasets\\'

In [3]:
standardized_df = pd.read_pickle(reldir + 'combined.csv', compression='zip')

In [4]:
master_df = pd.read_pickle(reldir + 'songdata.csv', compression='zip')

In [5]:
standardized_df.head()

user                song  count  \
0  b80344d063b5ccb3212f76538f3d9e43d87dca9e  SOWEZSI12A81C21CE6      1   
1  833c530ecda3d99deb8395f70400aa3999783d91  SOWEZSI12A81C21CE6      2   
2  d6c5bd2b570b4faf8964d7ed04f3392ff505d2be  SOWEZSI12A81C21CE6      1   
3  724534729c9f5dc72a009269c2c225883e4775d2  SOWEZSI12A81C21CE6      1   
4  ee7aa84c164038c963cfd02a7e52a5598aa470c3  SOWEZSI12A81C21CE6      2   

  analysis_songs_track_id metadata_songs_artist_id metadata_songs_song_id  \
0      TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   
1      TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   
2      TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   
3      TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   
4      TRAUCNU128F42671EB       AR2UQQ51187B9AC816     SOWEZSI12A81C21CE6   

   analysis_songs_key  analysis_songs_key_confidence  analysis_songs_mode  \
0                   5                          0.351                    0   
1                   5                          0.351                    0   
2                   5                          0.351                    0   
3                   5                          0.351                    0   
4                   5                          0.351                    0   

   analysis_songs_mode_confidence  ...  tempo_binned  analysis_songs_duration  \
0                           0.318  ...             8                -0.280898   
1                           0.318  ...             8                -0.280898   
2                           0.318  ...             8                -0.280898   
3                           0.318  ...             8                -0.280898   
4                           0.318  ...             8                -0.280898   

   metadata_songs_artist_familiarity  metadata_songs_artist_hotttnesss  \
0                           0.732042                          1.243373   
1                           0.732042                          1.243373   
2                           0.732042                          1.243373   
3                           0.732042                          1.243373   
4                           0.732042                          1.243373   

   metadata_songs_song_hotttnesss  \
0                         5.09644   
1                         5.09644   
2                         5.09644   
3                         5.09644   
4                         5.09644   

                               metadata_artist_terms  \
0  b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   
1  b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   
2  b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   
3  b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   
4  b'flamenco'|b'soundtrack'|b'folk'|b'spanish'|b...   

   metadata_songs_artist_name  metadata_songs_release  metadata_songs_title  \
0                 Gipsy Kings           Greatest Hits     Tu Quieres Volver   
1                 Gipsy Kings           Greatest Hits     Tu Quieres Volver   
2                 Gipsy Kings           Greatest Hits     Tu Quieres Volver   
3                 Gipsy Kings           Greatest Hits     Tu Quieres Volver   
4                 Gipsy Kings           Greatest Hits     Tu Quieres Volver   

         musicbrainz_artist_mbtags  
0  b'classic pop and rock'|b'folk'  
1  b'classic pop and rock'|b'folk'  
2  b'classic pop and rock'|b'folk'  
3  b'classic pop and rock'|b'folk'  
4  b'classic pop and rock'|b'folk'  

[5 rows x 27 columns]

https://jessesw.com/Rec-System/

https://towardsdatascience.com/building-a-collaborative-filtering-recommender-system-with-clickstream-data-dffc86c8c65

## Attempt 1 Using jessesw.com

In [6]:
## Check that all counts in dataset are 1 or greater
standardized_df.describe()

count  analysis_songs_key  analysis_songs_key_confidence  \
count  772661.000000       772661.000000                  772661.000000   
mean        2.684340            5.673522                       0.432972   
std         5.454646            3.615617                       0.270784   
min         1.000000            0.000000                       0.000000   
25%         1.000000            2.000000                       0.222000   
50%         1.000000            6.000000                       0.437000   
75%         2.000000            9.000000                       0.630000   
max       771.000000           11.000000                       1.000000   

       analysis_songs_mode  analysis_songs_mode_confidence  \
count        772661.000000                   772661.000000   
mean              0.637005                        0.489415   
std               0.480864                        0.165614   
min               0.000000                        0.000000   
25%               0.000000                        0.397000   
50%               1.000000                        0.480000   
75%               1.000000                        0.596000   
max               1.000000                        1.000000   

       analysis_songs_tempo  analysis_songs_time_signature  \
count         772661.000000                  772661.000000   
mean             126.560860                       3.730847   
std               31.789887                       1.021593   
min                0.000000                       0.000000   
25%              102.984000                       4.000000   
50%              121.884000                       4.000000   
75%              146.022000                       4.000000   
max              253.357000                       7.000000   

       analysis_songs_time_signature_confidence  musicbrainz_songs_year  \
count                             772661.000000           772661.000000   
mean                                   0.626652             1543.714245   
std                                    0.361508              839.540567   
min                                    0.000000                0.000000   
25%                                    0.358000             1977.000000   
50%                                    0.742000             2000.000000   
75%                                    0.994000             2006.000000   
max                                    1.000000             2010.000000   

        years_binned   tempo_binned  analysis_songs_duration  \
count  772661.000000  772661.000000            772661.000000   
mean        2.646629       4.783883                 0.158297   
std         1.620828       2.637524                 0.801748   
min         0.000000       0.000000                -2.151904   
25%         1.000000       3.000000                -0.228571   
50%         3.000000       5.000000                 0.052587   
75%         4.000000       7.000000                 0.464172   
max         4.000000       9.000000                14.586918   

       metadata_songs_artist_familiarity  metadata_songs_artist_hotttnesss  \
count                      772661.000000                     772661.000000   
mean                            1.137810                          1.828713   
std                             0.594691                          1.180343   
min                            -2.813196                         -2.960846   
25%                             0.847774                          1.077402   
50%                             1.298247                          1.707256   
75%                             1.572213                          2.589013   
max                             2.176964                          5.457244   

       metadata_songs_song_hotttnesss  
count                   772661.000000  
mean                         3.760755  
std                          2.317352  
min                         -4.007280  
25%                          2.414042  
50%                      

In [7]:
## Get our unique users
users = list(np.sort(standardized_df.user.unique()))
## Get our unique songs
songs = list(standardized_df.song.unique())
## Get counts of songs listened to
song_count = list(standardized_df['count'])

In [8]:
import scipy.sparse as sparse
from scipy.sparse.linalg import spsolve

## Get associated row indices
rows = pd.Categorical(standardized_df.user, categories=users).codes
## Get associated column indices
cols = pd.Categorical(standardized_df.song, categories=songs).codes
matrix = sparse.csr_matrix((song_count, (rows, cols)), shape=(len(users), len(songs)))

In [9]:
matrix

<418252x3675 sparse matrix of type '<class 'numpy.intc'>'
	with 772661 stored elements in Compressed Sparse Row format>

We have a total of 418,252 users and 3,675 songs, with 772,661 songs heard.

In [10]:
## Calculate matrix sparsity
# Obtain number of possible interactions in the matrix
matrix_size = matrix.shape[0]*matrix.shape[1]
# Obtain number of songs interacted with
num_songs_heard = len(matrix.nonzero()[0])
sparsity = 100*(1-(num_songs_heard/matrix_size))
sparsity

99.94973176669653

99.9% of the interaction matrix is sparse whereas the maximum that usually works for collaborative filtering is 99.5%. Therefore, we can attempt to build a matrix factorization model, but this may not perform well.

In [14]:
# Install a pip package in the current Jupyter kernel
import sys
!{sys.executable} -m pip install implicit

  Using cached implicit-0.4.4.tar.gz (1.1 MB)
  Created wheel for implicit: filename=implicit-0.4.4-cp38-cp38-win_amd64.whl size=612867 sha256=0c525b2fc8cd2b45d8a245a88d932f1729d208e904b55c4a8507e8c005aa46f0
  Stored in directory: c:\users\shefa\appdata\local\pip\cache\wheels\00\ac\67\6f4536c819ed560c2c7e17c0f7a920e3e50c26108616087d05
Successfully built implicit


## Attempt 2 Using Towards Data Science

In [11]:
#content_matrix = sparse.csr_matrix((song_count, (rows, cols)), shape=(len(songs), len(users)))

import scipy.sparse as sparse
from scipy.sparse.linalg import spsolve
from scipy.sparse import csr_matrix

standardized_df['user']=standardized_df['user'].astype("category")
standardized_df['song']=standardized_df['song'].astype("category")
standardized_df['user_id']= standardized_df['user'].cat.codes
standardized_df['song_id']=standardized_df['song'].cat.codes

sparse_content_user = sparse.csr_matrix((standardized_df['count'].astype(float), (standardized_df['song_id'], standardized_df['user_id'])))
sparse_user_content = sparse.csr_matrix((standardized_df['count'].astype(float), (standardized_df['user_id'], standardized_df['song_id'])))

In [28]:
import implicit
## will need to tweak the number of latent factors by potentially identifying loss function stability
## will need to tweak regularization to tune parameters
## user cross validation to tune the optimum factors in model
## Reasonable solution usually occurs in 20 iterations or less
## Knowing the number of latent or hidden factors influencing the model is impossible. We should
## start at 5-10, then increase by 5 until performance of model stops improving
model = implicit.als.AlternatingLeastSquares(factors=10, regularization=0.1, iterations=20)
#model can be evaluated by selecting model that generates the lowest RMSE on training set.

Link to help explain confidence and alpha's impact on confidence:
https://stats.stackexchange.com/questions/155296/alternating-least-squares-als-why-two-different-kinds-of-setting-for-lambd

In [29]:
## The paper used to develop ALS used alpha of 40
alpha = 40
data = (sparse_content_user*alpha).astype('double')
model.fit(data)

  0%|          | 0/20 [00:00<?, ?it/s]

In [30]:
## Obtain recommendations for a given user
user_id = 350627
recommended = model.recommend(user_id, sparse_user_content, N=5)
print(recommended)

[(693, 0.79169756), (1722, 0.71202666), (203, 0.70730686), (2730, 0.6374319), (2942, 0.63132167)]


In [31]:
## Get similar items

song_id=2918
n_similar = 3
similar = model.similar_items(song_id, n_similar)
print(similar)

[(2918, 0.99999994), (162, 0.8860025), (227, 0.8719206)]


https://campus.datacamp.com/courses/recommendation-engines-in-pyspark/what-if-you-dont-have-customer-ratings?ex=6

In [ ]:
#Use ROEM as evaluation metric